libvast/vast/system/evaluator.hpp

```diff
@@ -0,0 +1,80 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+#include <caf/actor.hpp>
+#include <caf/fwd.hpp>
+#include <caf/typed_response_promise.hpp>
+
+#include "vast/expression.hpp"
+#include "vast/ids.hpp"
+
+namespace vast::system {
+
+/// @relates evaluator
+struct evaluator_state {
+  using sub_hits_map = std::unordered_map<predicate, std::pair<size_t, ids>>;
+
+  evaluator_state(caf::event_based_actor* self);
+
+  void init(caf::actor client, expression expr);
+
+  /// Updates `sub_hits` and may triggers re-evaluation of the expression tree.
+  void handle_indexer_result(const predicate& pred, const ids& result);
+
+  /// Updates `sub_hits` and may triggers re-evaluation of the expression tree.
+  void handle_missing_indexer_result(const predicate& pred,
+                                     const caf::error& err);
+
+  /// Evaluates the predicate-tree and may produces new deltas.
+  void update();
+
+  /// Decrements the `pending_responses` and sends 'done' to the client when it
+  /// reaches 0.
+  void decrement_pending();
+
+  /// Returns the `sub_hits` entry for `pred` or `nullptr`.
+  sub_hits_map::mapped_type* sub_hits_of(const predicate& pred);
```

Comment:
How about calling this simply `hits_for`? Based on the types, it sounds more natural: you would get the hits for a given predicate.